File: Tellma/Entities/ReportDefinition.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System;

namespace Tellma.Entities
{
    [StrongEntity]
    [EntityDisplay(Singular = "ReportDefinition", Plural = "ReportDefinitions")]
    public class ReportDefinitionForSave<TParameter, TRow, TColumn, TMeasure, TSelect, TRole> : EntityWithKey<int>
    {
        [Display(Name = "Code")]
        [NotNull]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Code { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Primary)]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Title { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Secondary)]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Title2 { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Ternary)]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Title3 { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Primary)]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string Description { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Secondary)]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string Description2 { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Ternary)]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string Description3 { get; set; }

        [Display(Name = "ReportDefinition_Type")]
        [Required]
        [NotNull]
        [AlwaysAccessible]
        [ChoiceList(new object[] { "Summary", "Details" },
            new string[] { "ReportDefinition_Type_Summary", "ReportDefinition_Type_Details" })]
        public string Type { get; set; }

        [Display(Name = "ReportDefinition_Chart")]
        [AlwaysAccessible]
        [ChoiceList(new object[] {
                // 0 Dimensions
                "Card",
                // 1 Dimension
                "BarsVertical", "BarsHorizontal", "Pie", "Doughnut", "TreeMap", "NumberCards", "Gauge",
                // 1 or 2 Dimensions
                "Line", "Area",
                // 2 Dimensions
                "BarsVerticalGrouped", "BarsVerticalStacked", "BarsVerticalNormalized", "BarsHorizontalGrouped",
                "BarsHorizontalStacked", "BarsHorizontalNormalized", "HeatMap"
            },
            new string[] {
                // 0 Dimensions
                "ReportDefinition_Chart_Card",
                // 1 Dimension
                "ReportDefinition_Chart_BarsVertical", "ReportDefinition_Chart_BarsHorizontal", "ReportDefinition_Chart_Pie", "ReportDefinition_Chart_Doughnut", "ReportDefinition_Chart_TreeMap", "ReportDefinition_Chart_NumberCards", "ReportDefinition_Chart_Gauge",
                // 1 or 2 Dimensions
                "ReportDefinition_Chart_Line", "ReportDefinition_Chart_Area",
                // 2 Dimensions
                "ReportDefinition_Chart_BarsVerticalGrouped", "ReportDefinition_Chart_BarsVerticalStacked", "ReportDefinition_Chart_BarsVerticalNormalized", "ReportDefinition_Chart_BarsHorizontalGrouped",
                "ReportDefinition_Chart_BarsHorizontalStacked", "ReportDefinition_Chart_BarsHorizontalNormalized", "ReportDefinition_Chart_HeatMap"
            })]
        public string Chart { get; set; }

        [Display(Name = "ReportDefinition_DefaultsToChart")]
        [AlwaysAccessible]
        public bool? DefaultsToChart { get; set; }

        [Display(Name = "ReportDefinition_ChartOptions")]
        [AlwaysAccessible]
        [StringLength(1024)]
        public string ChartOptions { get; set; }

        [Display(Name = "ReportDefinition_Collection")]
        [Required]
        [NotNull]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Collection { get; set; }

        [Display(Name = "ReportDefinition_DefinitionId")]
        [AlwaysAccessible]
        public int? DefinitionId { get; set; }

        [Display(Name = "ReportDefinition_Filter")]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string Filter { get; set; } // On drill down for summary

        [Display(Name = "ReportDefinition_Having")]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string Having { get; set; } // On drill down for summary

        [Display(Name = "ReportDefinition_OrderBy")]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string OrderBy { get; set; } // On drill down for summary

        [Display(Name = "ReportDefinition_Top")]
        [AlwaysAccessible]
        public int? Top { get; set; }

        [Display(Name = "ReportDefinition_ShowColumnsTotal")]
        [AlwaysAccessible]
        public bool? ShowColumnsTotal { get; set; }

        [MultilingualDisplay(Name = "ReportDefinition_ColumnsTotalLabel", Language = Language.Primary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string ColumnsTotalLabel { get; set; }

        [MultilingualDisplay(Name = "ReportDefinition_ColumnsTotalLabel", Language = Language.Secondary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string ColumnsTotalLabel2 { get; set; }

        [MultilingualDisplay(Name = "ReportDefinition_ColumnsTotalLabel", Language = Language.Ternary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string ColumnsTotalLabel3 { get; set; }

        [Display(Name = "ReportDefinition_ShowRowsTotal")]
        [AlwaysAccessible]
        public bool? ShowRowsTotal { get; set; }

        [MultilingualDisplay(Name = "ReportDefinition_RowsTotalLabel", Language = Language.Primary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string RowsTotalLabel { get; set; }

        [MultilingualDisplay(Name = "ReportDefinition_RowsTotalLabel", Language = Language.Secondary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string RowsTotalLabel2 { get; set; }

        [MultilingualDisplay(Name = "ReportDefinition_RowsTotalLabel", Language = Language.Ternary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string RowsTotalLabel3 { get; set; }

        [Display(Name = "ReportDefinition_IsCustomDrilldown")]
        [AlwaysAccessible]
        public bool? IsCustomDrilldown { get; set; }

        [Display(Name = "Definition_ShowInMainMenu")]
        [AlwaysAccessible]
        public bool? ShowInMainMenu { get; set; }

        [Display(Name = "MainMenuSection")]
        [ChoiceList(new object[] { "Mail", "Financials", "Cash", "FixedAssets", "Inventory", "Production", "Purchasing", "Marketing", "Sales", "HumanCapital", "Investments", "Maintenance", "Administration", "Security", "Studio", "Help" },
            new string[] { "Menu_Mail", "Menu_Financials", "Menu_Cash", "Menu_FixedAssets", "Menu_Inventory", "Menu_Production", "Menu_Purchasing", "Menu_Marketing", "Menu_Sales", "Menu_HumanCapital", "Menu_Investments", "Menu_Maintenance", "Menu_Administration", "Menu_Security", "Menu_Studio", "Menu_Help" })]
        [AlwaysAccessible]
        public string MainMenuSection { get; set; }

        [Display(Name = "MainMenuIcon")]
       // [ChoiceList(new object[] { "clipboard", "chart-pie" })]
        [AlwaysAccessible]
        public string MainMenuIcon { get; set; }

        [Display(Name = "MainMenuSortKey")]
        [AlwaysAccessible]
        public decimal? MainMenuSortKey { get; set; }

        [Display(Name = "ReportDefinition_Parameters")]
        [ForeignKey(nameof(ReportDefinitionParameter.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TParameter> Parameters { get; set; }

        [Display(Name = "ReportDefinition_Rows")]
        [ForeignKey(nameof(ReportDefinitionRow.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TRow> Rows { get; set; }

        [Display(Name = "ReportDefinition_Columns")]
        [ForeignKey(nameof(ReportDefinitionColumn.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TColumn> Columns { get; set; }

        [Display(Name = "ReportDefinition_Measures")]
        [ForeignKey(nameof(ReportDefinitionMeasure.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TMeasure> Measures { get; set; }

        [Display(Name = "ReportDefinition_Select")]
        [ForeignKey(nameof(ReportDefinitionSelect.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TSelect> Select { get; set; }

        [Display(Name = "Definition_Roles")]
        [ForeignKey(nameof(ReportDefinitionRole.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TRole> Roles { get; set; }
    }

    public class ReportDefinitionForSave : ReportDefinitionForSave<ReportDefinitionParameterForSave, ReportDefinitionRowForSave, ReportDefinitionColumnForSave, ReportDefinitionMeasureForSave, ReportDefinitionSelectForSave, ReportDefinitionRoleForSave>
    {

    }

    public class ReportDefinition : ReportDefinitionForSave<ReportDefinitionParameter, ReportDefinitionRow, ReportDefinitionColumn, ReportDefinitionMeasure, ReportDefinitionSelect, ReportDefinitionRole>
    {
        [Display(Name = "CreatedAt")]
        [NotNull]
        public DateTimeOffset? CreatedAt { get; set; }

        [Display(Name = "CreatedBy")]
        [NotNull]
        public int? CreatedById { get; set; }

        [Display(Name = "ModifiedAt")]
        [NotNull]
        public DateTimeOffset? ModifiedAt { get; set; }

        [Display(Name = "ModifiedBy")]
        [NotNull]
        public int? ModifiedById { get; set; }

        // For Query

        [Display(Name = "CreatedBy")]
        [ForeignKey(nameof(CreatedById))]
        public User CreatedBy { get; set; }

        [Display(Name = "ModifiedBy")]
        [ForeignKey(nameof(ModifiedById))]
        public User ModifiedBy { get; set; }
    }

    public class ReportDefinitionParameterForSave : EntityWithKey<int>
    {
        [Display(Name = "Parameter_Key")]
        [Required]
        [NotNull]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Key { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Primary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Secondary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label2 { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Ternary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label3 { get; set; }

        [Display(Name = "Parameter_Visibility")]
        [AlwaysAccessible]
        [VisibilityChoiceList]
        public string Visibility { get; set; }

        [Display(Name = "ReportDefinition_DefaultExpression")]
        [StringLength(255)]
        [AlwaysAccessible]
        public string DefaultExpression { get; set; }

        [Display(Name = "Definition_Control")]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Control { get; set; }

        [Display(Name = "Definition_ControlOptions")]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string ControlOptions { get; set; }
    }

    public class ReportDefinitionParameter : ReportDefinitionParameterForSave
    {
        [NotNull]
        public int? ReportDefinitionId { get; set; }

        [NotNull]
        public int? Index { get; set; }
    }

    public class ReportDefinitionSelectForSave : EntityWithKey<int>
    {
        [Display(Name = "ReportDefinition_Expression")]
        [Required]
        [NotNull]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Expression { get; set; }

        [Display(Name = "ReportDefinition_Localize")]
        [NotNull]
        public bool? Localize { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Primary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Secondary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label2 { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Ternary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label3 { get; set; }

        [Display(Name = "Definition_Control")]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Control { get; set; }

        [Display(Name = "Definition_ControlOptions")]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string ControlOptions { get; set; }
    }

    public class ReportDefinitionSelect : ReportDefinitionSelectForSave
    {
        [NotNull]
        public int? ReportDefinitionId { get; set; }

        [NotNull]
        public int? Index { get; set; }
    }

    public abstract class ReportDefinitionDimension<TAttribute> : EntityWithKey<int>
    {
        [Display(Name = "ReportDefinition_KeyExpression")]
        [Required]
        [NotNull]
        [StringLength(255)]
        [AlwaysAccessible]
        public string KeyExpression { get; set; }

        [Display(Name = "ReportDefinition_DisplayExpression")]
        [StringLength(255)]
        [AlwaysAccessible]
        public string DisplayExpression { get; set; }

        [Display(Name = "ReportDefinition_Localize")]
        [NotNull]
        public bool? Localize { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Primary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Secondary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label2 { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Ternary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label3 { get; set; }

        [Display(Name = "ReportDefinition_OrderDirection")]
        [AlwaysAccessible]
        [ChoiceList(new object[] { "asc", "desc" },
            new string[] { "ReportDefinition_OrderDirection_Asc", "ReportDefinition_OrderDirection_Desc" })]
        public string OrderDirection { get; set; }

        [Display(Name = "ReportDefinition_AutoExpand")]
        [AlwaysAccessible]
        public int? AutoExpandLevel { get; set; }

        [Display(Name = "ReportDefinition_ShowAsTree")]
        [NotNull]
        [AlwaysAccessible]
        public bool? ShowAsTree { get; set; }

        [Display(Name = "Definition_Control")]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Control { get; set; }

        [Display(Name = "Definition_ControlOptions")]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string ControlOptions { get; set; }

        [Display(Name = "ReportDefinition_Attributes")]
        [ForeignKey(nameof(ReportDefinitionDimensionAttribute.ReportDefinitionDimensionId))]
        [AlwaysAccessible]
        public List<TAttribute> Attributes { get; set; }
    }

    public class ReportDefinitionColumnForSave : ReportDefinitionDimension<ReportDefinitionDimensionAttributeForSave>
    {

    }

    public class ReportDefinitionColumn : ReportDefinitionDimension<ReportDefinitionDimensionAttribute>
    {
        [NotNull]
        public int? ReportDefinitionId { get; set; }

        [NotNull]
        public int? Index { get; set; }
    }

    public class ReportDefinitionRowForSave : ReportDefinitionDimension<ReportDefinitionDimensionAttributeForSave>
    {

    }

    public class ReportDefinitionRow : ReportDefinitionDimension<ReportDefinitionDimensionAttribute>
    {
        [NotNull]
        public int? ReportDefinitionId { get; set; }

        [NotNull]
        public int? Index { get; set; }
    }

    public class ReportDefinitionDimensionAttributeForSave : EntityWithKey<int>
    {
        [Display(Name = "ReportDefinition_Expression")]
        [NotNull]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Expression { get; set; }

        [Display(Name = "ReportDefinition_Localize")]
        [NotNull]
        public bool? Localize { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Primary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Secondary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label2 { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Ternary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label3 { get; set; }

        [Display(Name = "ReportDefinition_OrderDirection")]
        [AlwaysAccessible]
        [ChoiceList(new object[] { "asc", "desc" },
            new string[] { "ReportDefinition_OrderDirection_Asc", "ReportDefinition_OrderDirection_Desc" })]
        public string OrderDirection { get; set; }
    }

    public class ReportDefinitionDimensionAttribute : ReportDefinitionDimensionAttributeForSave
    {
        [NotNull]
        public int? ReportDefinitionDimensionId { get; set; }

        [NotNull]
        public int? Index { get; set; }
    }

    public class ReportDefinitionMeasureForSave : EntityWithKey<int>
    {
        [Display(Name = "ReportDefinition_Expression")]
        [Required]
        [NotNull]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Expression { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Primary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Secondary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label2 { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Ternary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Label3 { get; set; }

        [Display(Name = "ReportDefinition_OrderDirection")]
        [AlwaysAccessible]
        [ChoiceList(new object[] { "asc", "desc" },
            new string[] { "ReportDefinition_OrderDirection_asc", "ReportDefinition_OrderDirection_desc" })]
        public string OrderDirection { get; set; }

        [Display(Name = "Definition_Control")]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Control { get; set; }

        [Display(Name = "Definition_ControlOptions")]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string ControlOptions { get; set; }

        [Display(Name = "ReportDefinition_DangerWhen")]
        [StringLength(255)]
        [AlwaysAccessible]
        public string DangerWhen { get; set; }

        [Display(Name = "ReportDefinition_WarningWhen")]
        [StringLength(255)]
        [AlwaysAccessible]
        public string WarningWhen { get; set; }

        [Display(Name = "ReportDefinition_SuccessWhen")]
        [StringLength(255)]
        [AlwaysAccessible]
        public string SuccessWhen { get; set; }
    }

    public class ReportDefinitionMeasure : ReportDefinitionMeasureForSave
    {
        [NotNull]
        public int? ReportDefinitionId { get; set; }

        [NotNull]
        public int? Index { get; set; }
    }

    public class ReportDefinitionRoleForSave : EntityWithKey<int>
    {
        [Display(Name = "Definition_Role")]
        [NotNull]
        public int? RoleId { get; set; }
    }

    public class ReportDefinitionRole : ReportDefinitionRoleForSave
    {
        [NotNull]
        public int? ReportDefinitionId { get; set; }

        [Display(Name = "Definition_Role")]
        [ForeignKey(nameof(RoleId))]
        public Role Role { get; set; }
    }
}

File: Tellma/ClientApp/src/app/data/entities/report-definition.ts
// tslint:disable:variable-name
// tslint:disable:max-line-length
import { EntityForSave } from './base/entity-for-save';
import { WorkspaceService } from '../workspace.service';
import { TranslateService } from '@ngx-translate/core';
import { SettingsForClient } from '../dto/settings-for-client';
import { Collection, Control, EntityDescriptor } from './base/metadata';
import { EntityWithKey } from './base/entity-with-key';
import { DefinitionVisibility as Visibility, MainMenuSection, MainMenuIcon, mainMenuSectionPropDescriptor, mainMenuIconPropDescriptor, mainMenuSortKeyPropDescriptor } from './base/definition-common';

export type ReportOrderDirection = 'asc' | 'desc';
export type ReportType = 'Summary' | 'Details';
export type ChartType = 'Card' | 'BarsVertical' | 'BarsVerticalGrouped' | 'BarsVerticalStacked' |
    'BarsVerticalNormalized' | 'BarsHorizontal' | 'BarsHorizontalGrouped' | 'BarsHorizontalStacked' |
    'BarsHorizontalNormalized' | 'Line' | 'Area' | 'AreaStacked' | 'AreaNormalized' | 'Pie' | 'Doughnut' |
    'HeatMap' | 'TreeMap' | 'NumberCards' | 'Gauge' | 'Radar';

export interface ReportDefinitionForSave<
    TParameter = ReportDefinitionParameterForSave,
    TRow = ReportDefinitionRowForSave,
    TColumn = ReportDefinitionColumnForSave,
    TMeasure = ReportDefinitionMeasureForSave,
    TSelect = ReportDefinitionSelectForSave> extends EntityForSave {

    Title?: string;
    Title2?: string;
    Title3?: string;
    Description?: string;
    Description2?: string;
    Description3?: string;
    Type?: ReportType; // summary or details
    Chart?: ChartType;
    DefaultsToChart?: boolean;
    ChartOptions?: string; // JSON
    Collection?: Collection;
    DefinitionId?: number;
    Filter?: string;
    Having?: string;
    OrderBy?: string;
    Top?: number;
    ShowColumnsTotal?: boolean;
    ColumnsTotalLabel?: string;
    ColumnsTotalLabel2?: string;
    ColumnsTotalLabel3?: string;
    ShowRowsTotal?: boolean;
    RowsTotalLabel?: string;
    RowsTotalLabel2?: string;
    RowsTotalLabel3?: string;
    IsCustomDrilldown?: boolean;
    MainMenuSection?: MainMenuSection;
    MainMenuIcon?: MainMenuIcon;
    MainMenuSortKey?: number;
    ShowInMainMenu?: boolean;

    Select?: TSelect[];
    Parameters?: TParameter[];
    Rows?: TRow[];
    Columns?: TColumn[];
    Measures?: TMeasure[];
}

export interface ReportDefinition extends ReportDefinitionForSave<
    ReportDefinitionParameter, ReportDefinitionRow, ReportDefinitionColumn, ReportDefinitionMeasure, ReportDefinitionSelect> {
    CreatedAt?: string;
    CreatedById?: number | string;
    ModifiedAt?: string;
    ModifiedById?: number | string;
}

export interface ReportDefinitionParameterForSave extends EntityForSave {
    Key?: string; // e.g. 'FromDate'
    Label?: string;
    Label2?: string;
    Label3?: string;
    Visibility?: Visibility;
    DefaultExpression?: string;
    Control?: Control;
    ControlOptions?: string; // JSON
}

// tslint:disable-next-line:no-empty-interface
export interface ReportDefinitionParameter extends ReportDefinitionParameterForSave {
    ReportDefinitionId?: number;
}

export interface ReportDefinitionSelectForSave extends EntityForSave {
    Expression?: string;
    Localize?: boolean;
    Label?: string;
    Label2?: string;
    Label3?: string;
    Control?: Control;
    ControlOptions?: string; // JSON
}

// tslint:disable-next-line:no-empty-interface
export interface ReportDefinitionSelect extends ReportDefinitionSelectForSave {
    ReportDefinitionId?: number;
}

export interface ReportDefinitionDimension<TAttribute> extends EntityForSave {
    KeyExpression?: string;
    DisplayExpression?: string;
    Localize?: boolean;
    Label?: string;
    Label2?: string;
    Label3?: string;
    OrderDirection?: ReportOrderDirection;
    AutoExpandLevel?: number;
    ShowAsTree?: boolean;
    Attributes?: TAttribute[];
}

// tslint:disable-next-line:no-empty-interface
export interface ReportDefinitionColumnForSave extends ReportDefinitionDimension<ReportDefinitionDimensionAttributeForSave> {
}

// tslint:disable-next-line:no-empty-interface
export interface ReportDefinitionColumn extends ReportDefinitionDimension<ReportDefinitionDimensionAttribute> {
    ReportDefinitionId?: number;
}

// tslint:disable-next-line:no-empty-interface
export interface ReportDefinitionRowForSave extends ReportDefinitionDimension<ReportDefinitionDimensionAttributeForSave> {
}

// tslint:disable-next-line:no-empty-interface
export interface ReportDefinitionRow extends ReportDefinitionDimension<ReportDefinitionDimensionAttribute> {
    ReportDefinitionId?: number;
}

export interface ReportDefinitionDimensionAttributeForSave extends EntityForSave {
    Expression?: string;
    Localize?: boolean;
    Label?: string;
    Label2?: string;
    Label3?: string;
    OrderDirection?: ReportOrderDirection;

}

export interface ReportDefinitionDimensionAttribute extends ReportDefinitionDimensionAttributeForSave {
    ReportDefinitionDimensionId?: number;
}


export interface ReportDefinitionMeasureForSave extends EntityForSave {
    Expression?: string;
    Label?: string;
    Label2?: string;
    Label3?: string;
    OrderDirection?: ReportOrderDirection;
    Control?: Control;
    ControlOptions?: string; // JSON
    DangerWhen?: string;
    WarningWhen?: string;
    SuccessWhen?: string;
}

// tslint:disable-next-line:no-empty-interface
export interface ReportDefinitionMeasure extends ReportDefinitionMeasureForSave {
    ReportDefinitionId?: number;
}

const _select = ['', '2', '3'].map(pf => 'Title' + pf);
let _settings: SettingsForClient;
let _cache: EntityDescriptor;

export function metadata_ReportDefinition(wss: WorkspaceService, trx: TranslateService): EntityDescriptor {
    const ws = wss.currentTenant;
    // Some global values affect the result, we check here if they have changed, otherwise we return the cached result
    if (ws.settings !== _settings) {
        _settings = ws.settings;
        _cache = {
            collection: 'ReportDefinition',
            titleSingular: () => trx.instant('ReportDefinition'),
            titlePlural: () => trx.instant('ReportDefinitions'),
            apiEndpoint: 'report-definitions',
            masterScreenUrl: 'report-definitions',
            select: _select,
            orderby: () => ws.isSecondaryLanguage ? [_select[1], _select[0]] : ws.isTernaryLanguage ? [_select[2], _select[0]] : [_select[0]],
            inactiveFilter: null, // TODO
            format: (item: EntityWithKey) => (ws.getMultilingualValueImmediate(item, _select[0]) || trx.instant('Untitled')),
            formatFromVals: (vals: any[]) => (ws.localize(vals[0], vals[1], vals[2]) || trx.instant('Untitled')),
            properties: {
                Id: { noSeparator: true, datatype: 'numeric', control: 'number', label: () => trx.instant('Id'), minDecimalPlaces: 0, maxDecimalPlaces: 0 },
                Title: { datatype: 'string', control: 'text', label: () => trx.instant('Title') + ws.primaryPostfix },
                Title2: { datatype: 'string', control: 'text', label: () => trx.instant('Title') + ws.secondaryPostfix },
                Title3: { datatype: 'string', control: 'text', label: () => trx.instant('Title') + ws.ternaryPostfix },
                Description: { datatype: 'string', control: 'text', label: () => trx.instant('Description') + ws.primaryPostfix },
                Description2: { datatype: 'string', control: 'text', label: () => trx.instant('Description') + ws.secondaryPostfix },
                Description3: { datatype: 'string', control: 'text', label: () => trx.instant('Description') + ws.ternaryPostfix },
                Type: {
                    datatype: 'string',
                    control: 'choice',
                    label: () => trx.instant('ReportDefinition_Type'),
                    choices: ['Summary', 'Details'],
                    format: (c: string) => trx.instant(`ReportDefinition_Type_${c}`)
                },
                Chart: {
                    datatype: 'string',
                    control: 'choice',
                    label: () => trx.instant('ReportDefinition_Chart'),
                    choices: [ // Ordered by number of supported dimensions for ease of selection
                        // 0 Dimensions
                        'Card',
                        // 1 Dimension
                        'BarsVertical', 'BarsHorizontal', 'Pie', 'Doughnut', 'TreeMap', 'NumberCards', 'Gauge',
                        // 1 or 2 Dimensions
                        'Line', 'Area',
                        // 2 Dimensions
                        'BarsVerticalGrouped', 'BarsVerticalStacked', 'BarsVerticalNormalized', 'BarsHorizontalGrouped',
                        'BarsHorizontalStacked', 'BarsHorizontalNormalized', /* 'AreaStacked', 'AreaNormalized', 'Radar', */ 'HeatMap'],
                    format: (c: string) => !!c ? trx.instant(`ReportDefinition_Chart_${c}`) : c
                },
                DefaultsToChart: { datatype: 'bit', control: 'check', label: () => trx.instant('ReportDefinition_DefaultsToChart') },
                ChartOptions: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_ChartOptions') },
                Collection: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_Collection') },
                DefinitionId: { noSeparator: true, datatype: 'numeric', control: 'number', label: () => trx.instant('ReportDefinition_DefinitionId'), minDecimalPlaces: 0, maxDecimalPlaces: 0 },
                Filter: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_Filter') },
                Having: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_Having') },
                OrderBy: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_OrderBy') },
                Top: { datatype: 'numeric', control: 'number', label: () => trx.instant('ReportDefinition_Top'), minDecimalPlaces: 0, maxDecimalPlaces: 0, noSeparator: false },
                ShowColumnsTotal: { datatype: 'bit', control: 'check', label: () => trx.instant('ReportDefinition_ShowColumnsTotal') },
                ColumnsTotalLabel: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_ColumnsTotalLabel') + ws.primaryPostfix },
                ColumnsTotalLabel2: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_ColumnsTotalLabel') + ws.secondaryPostfix },
                ColumnsTotalLabel3: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_ColumnsTotalLabel') + ws.ternaryPostfix },
                ShowRowsTotal: { datatype: 'bit', control: 'check', label: () => trx.instant('ReportDefinition_ShowRowsTotal') },
                RowsTotalLabel: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_RowsTotalLabel') + ws.primaryPostfix },
                RowsTotalLabel2: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_RowsTotalLabel') + ws.secondaryPostfix },
                RowsTotalLabel3: { datatype: 'string', control: 'text', label: () => trx.instant('ReportDefinition_RowsTotalLabel') + ws.ternaryPostfix },
                IsCustomDrilldown: { datatype: 'bit', control: 'check', label: () => trx.instant('ReportDefinition_IsCustomDrilldown') },
                MainMenuSection: mainMenuSectionPropDescriptor(trx),
                MainMenuIcon: mainMenuIconPropDescriptor(trx),
                MainMenuSortKey: mainMenuSortKeyPropDescriptor(trx),
                ShowInMainMenu: { datatype: 'bit', control: 'check', label: () => trx.instant('ReportDefinition_ShowInMainMenu') },
                CreatedAt: { datatype: 'datetimeoffset', control: 'datetime', label: () => trx.instant('CreatedAt') },
                CreatedBy: { datatype: 'entity', control: 'User', label: () => trx.instant('CreatedBy'), foreignKeyName: 'CreatedById' },
                ModifiedAt: { datatype: 'datetimeoffset', control: 'datetime', label: () => trx.instant('ModifiedAt') },
                ModifiedBy: { datatype: 'entity', control: 'User', label: () => trx.instant('ModifiedBy'), foreignKeyName: 'ModifiedById' }
            }
        };

        if (!ws.settings.SecondaryLanguageId) {
            delete _cache.properties.Title2;
            delete _cache.properties.Description2;
        }

        if (!ws.settings.TernaryLanguageId) {
            delete _cache.properties.Title3;
            delete _cache.properties.Description3;
        }
    }

    return _cache;
}
